Add --subdir argument to allow for go.mod files in different locations & numerous fixes

e4t opened the pull request:
Some Go projects may have sub-projects that are build independently. These come in subdirectories with their own go.mod. Allow to specify the directory where to search for a go.mod file.

While testing this there were numerous other issues which had to be addressed as well:
- the output vendor.tar.* file was not compressed correctly
- name of the top directory in the input tarball may not match tar-ball name.
- An empty --compression flag should just fall back to  DEFAULT_COMPRESSION.

aplanas commented on go_modules:
```diff
@@ -134,8 +134,28 @@ def extract(filename, dir):
+            s_head = subdir
+            s_match = False
+            while not s_match:
+                while True:
```
Is this loop equivalent to `os.path.basename()`

aplanas commented on go_modules:
```diff
@@ -156,15 +176,16 @@ def cmd_go_mod(cmd, dir):
 
-    log.info(f"Extracting {archive} to {outdir}")
-    extract(archive, outdir)
+    tempdir = tempfile.mkdtemp(dir=outdir, prefix='extract-')
```
Maybe `with tempfile.TemporaryDirectory() as tempdir:` is more helpfull here. I will create it in the correct place and will not change "outdir". Also will take care of removing the directory.

aplanas commented on go_modules:
```diff
@@ -176,13 +197,15 @@ def main():
         try:
-            to_remove = os.path.join(outdir, basename)
-            shutil.rmtree(to_remove)
+            shutil.rmtree(tempdir)
```
With the `with` you can remove all this `try` block.

e4t replied:
Sure, but I cannot figure out how to handle the error when the directory cannot be removed for some reason - as this exception would happen in a cleanup of the context manager.
@aplanas Any hint?

aplanas replied:
Reading the doc about [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) we can see that it will remove automatically the directory for you. If it fail, you will fail too.

Also inspecting the [code](https://github.com/python/cpython/blob/3.10/Lib/tempfile.py#L827) we can see that the clean method goes much much farther that the `rmtree` in your code, like adjusting permissions, behaving differently when an exception happens, etc. So IMHO just use the context manager with TemporaryDirectory and it will be cleaned automatically.

e4t commented on go_modules:
```diff
@@ -176,13 +197,15 @@ def main():
-        with tarfile.open(vendor_tarfile, "w:" + args.compression) as tar:
+        tarmode = "w:"
+        if args.compression != "tar":
+            tarmode += args.compression or DEFAULT_COMPRESSION
```
It contains the same logic - but it returns 'tar' + <compression_string>. Here only <compression_string> is needed.

e4t commented:
@aplanas - Thanks!
I've updated the code following your suggestions. Two patches were obsolete. I did read the documentation on `TemporaryDirectory`: this is where it talks about the exception generated on error during cleanup (unless `ignore_cleanup_errors` is set) - I hadn't looked at the code, though.
I've also simplified the changes to `find_file()` - I haven't found more useful `os.path` calls, however, the intentions should be a bit more clear, now.

aplanas commented:
@e4t thanks!

@jfkw Can you review the change?

e4t commented on go_modules:
```diff
@@ -187,7 +188,10 @@ def cmd_go_mod(cmd, dir):
     log.info(f"go mod {cmd}")
     # subprocess.run() returns CompletedProcess cp
-    cp = run(["go", "mod", cmd], cwd=dir, capture_output=True, text=True)
+    if (sys.version_info.major * 1000 + sys.version_info.minor > 3006):
```
Thanks! 
To cover the 'micro' version number as well, this should probably be 
`if sys.version_info >= (3, 7):`

e4t commented:
After chatting with Jeff I've decided to rework my patches again. 
Searching for the go.mod file may not render the correct result - especially if there are more such files around and the subdir mechanism used is not totally fool-proof. 
It seems to be better to look for it in the topmost directory only and expect the user to provide the directory explicitly if it is not here (or another one is to be used).

jfkw commented:
Thank you for the contributed fixes and the new `--subdir` argument. I have added one commit to make that argument optional, and tested it with a number of Go application packages with success.

These commits will be the next tagged release landing presently.

e4t commented:
> Thank you for the contributed fixes and the new `--subdir` argument. I have added one commit to make that argument optional, and tested it with a number of Go application packages with success.

Good point! Thank you!
> 
> These commits will be the next tagged release landing presently.

Thank you for the update!
